gui/rendering/ClippingPlane.cpp

```diff
@@ -0,0 +1,106 @@
+//
+// Created by hcaramizaru on 2/25/24.
+//
+
+#include "ClippingPlane.hpp"
+#include "ShaderProgram.hpp"
+#include "qopenglmousetracker.hpp"
+#include "sme/logger.hpp"
+#include <QVector3D>
+
+std::stack<uint32_t> generate(int32_t n) {
+
+  assert(n > 0);
+  std::stack<uint32_t> st;
+  for (int32_t i = n - 1; i >= 0; i--) {
+    st.push(i);
+  }
+  return st;
+}
+
+std::stack<uint32_t> rendering::ClippingPlane::available =
+    generate(MAX_NUMBER_PLANES);
+
+rendering::ClippingPlane::ClippingPlane(GLfloat a, GLfloat b, GLfloat c,
+                                        GLfloat d,
+                                        QOpenGLMouseTracker &OpenGLWidget,
+                                        bool active)
+    : a(a), b(b), c(c), d(d), status(active), OpenGLWidget(OpenGLWidget) {
+
+  assert(!available.empty());
+
+  this->OpenGLWidget.addClippingPlane(this);
+
+  planeIndex = available.top();
+  available.pop();
+}
+
+rendering::ClippingPlane::ClippingPlane(const QVector3D &normal,
+                                        const QVector3D &point,
+                                        QOpenGLMouseTracker &OpenGLWidget,
+                                        bool active)
+    : a(normal.x()), b(normal.y()), c(normal.z()),
+      d(-QVector3D::dotProduct(normal, point)), status(active),
+      OpenGLWidget(OpenGLWidget) {
+
+  assert(!available.empty());
+
+  this->OpenGLWidget.addClippingPlane(this);
+
+  planeIndex = available.top();
+  available.pop();
+}
+
+rendering::ClippingPlane::~ClippingPlane() {
+
+  Disable();
+  try {
+    available.push(planeIndex);
+  } catch (const std::exception &e) {
+    SPDLOG_ERROR("Exception caught: " + std::string(e.what()));
+  }
+  this->OpenGLWidget.deleteClippingPlane(this);
+}
+
+void rendering::ClippingPlane::SetClipPlane(GLfloat a, GLfloat b, GLfloat c,
+                                            GLfloat d) {
+  this->a = a;
+  this->b = b;
+  this->c = c;
+  this->d = d;
+}
+
+void rendering::ClippingPlane::SetClipPlane(QVector3D normal,
+                                            const QVector3D &point) {
+
+  normal.normalize();
+
+  SetClipPlane(normal.x(), normal.y(), normal.z(),
+               -QVector3D::dotProduct(normal, point));
+}
+
+void rendering::ClippingPlane::TranslateClipPlane(GLfloat value) {
+
+  GLfloat length = QVector3D(a, b, c).length();
+  QVector3D normal(a / length, b / length, c / length);
+
+  GLfloat p = d / length;
```

Comment:
isn't length=1 here?

Reply:
you can have normals of different size than 1.

Reply:
No necessary... you can have valid planes like this:  (a,b,c,d) = (6,7,8,1)